realtime: Blog for Realtime Broadcast and Presence Authz (#27691)
supabase/supabase: apps/www/_blog/2024-07-01-supabase-realtime-broadcast-and-presence-authorization.mdx

description: 'Create Authorization policies for Realtime Broadcast and Presence'
author: filipe
image: 2024-07-01-supabase-realtime-broadcast-and-presence-authorization/OG.png
thumb: 2024-07-01-supabase-realtime-broadcast-and-presence-authorization/thumb.png
categories:
tags:
  - realtime
  - engineering
date: '2024-07-01'
toc_depth: 2
---

Today we're releasing Realtime Authorization for Broadcast and Presence as Public Alpha.

For context, Supabase includes three useful extensions for building real-time applications.

1. [Broadcast](https://supabase.com/docs/guides/realtime/broadcast): Send ephemeral, low-latency messages between users.
2. [Presence](https://supabase.com/docs/guides/realtime/presence): Show when users are online and share state between users.
3. [Postgres Changes](https://supabase.com/docs/guides/realtime/postgres-changes): Listen to Postgres database changes.

This release introduces authorization for Broadcast and Presence using Row Level Security policies:

```sql
CREATE POLICY "authenticated user can listen to all messages"
ON "realtime"."messages"
AS PERMISSIVE FOR SELECT
TO authenticated
USING ( true );
```

> This follows a similar approach to [Supabase Storage](https://supabase.com/docs/guides/storage/security/access-control) where we create a new table in the `realtime` schema that represents your [Channels](https://supabase.com/docs/guides/realtime/concepts#channels) as objects so you can write policies where SELECT and INSERT will determines what your users can access and use from Realtime.

Reviewer: ```suggestion
The `realtime.messages` table is created automatically by Supabase Realtime. It represents your [Channels](/docs/guides/realtime/concepts#channels) as objects against which you can write RLS policies. We test for a user's ability to `SELECT` and `INSERT` into this table to determine whether they can join, read, or post to a Realtime Channel.

This approach of using RLS policies to control access should be familiar to users of [Supabase Storage](/docs/guides/storage/security/access-control).
```

## How it works without Authorization

When you want to connect to a Realtime Channel, you do something close to the following code:

```tsx
// With authenticated user token
client.realtime.setAuth(token)

client.channel('channel').subscribe((status: string, err: any) => {
  status === 'SUBSCRIBED' && console.log('Connected')
})
```

This will enable you to send messages but they will all be public and every user would be able to intercept them if they had your anon token.

> 

Reviewer: ```suggestion
## Adding Authorization to Realtime Channels

```

To convert this into a Realtime Channel that checks RLS policies requires only two steps:

- [Create RLS policies](#create-rls-policies)
- [Enabling Authorization on Channel subscription](#enabling-authorization-on-channel-subscription)

### Create RLS policies

Any RLS policy will work. Reference our [Github Discussion](github.com/orgs/supabase/discussions/22484) for more complex use cases.

With this example we'll keep it simple and allow authenticated users to connect to a channel, send and receive broadcast messages and track presences.

```sql
CREATE POLICY "authenticated user listen to all"
ON "realtime"."messages"
AS PERMISSIVE FOR SELECT
TO authenticated
USING ( true );

CREATE POLICY "authenticated user write to all"
ON "realtime"."messages"
AS PERMISSIVE FOR INSERT
TO authenticated
WITH CHECK ( true );
```

We also have a new database function called`realtime.topic()` which enables you to write Policies that take into consideration the name of the Channel trying to run the RLS Policy.

```sql
create policy "authenticated users can only read from 'locked' topic"
on "realtime"."messages"
as PERMISSIVE for SELECT
to authenticated
using ( realtime.topic() = 'locked' );
```

> ## Enabling Authorization on Channel subscription

Reviewer: ```suggestion
### Enabling Authorization on Channel subscription
```

> Now that you have your RLS policies set you will need to define on your Channel subscription that you intend to use a private channel. To achieve it we've introduced a new configuration parameter `private` to signal to Realtime servers that you want to check authorization on the channel.

Reviewer: ```suggestion
Now that you have your RLS policies set, you will need to define on your Channel subscription that you intend to use a private channel. To achieve it we've introduced a new configuration parameter `private` to signal to Realtime servers that you want to check authorization on the channel.
```

If you try to subscribe with an unauthorized user you will get a new error message informing the user that they do not have permission to access the topic.

```tsx
// With anon user
supabase
    .realtime

>     .channel("locked", { config: { private: true })

Reviewer: ```suggestion
    .channel("locked", { config: { private: true } })
```

Out of curiosity, what happens if you try to join a private channel without passing `private: true`? Assuming you get a similar error? 

Not super important, but it was the first question that popped to mind while reading this, so might be worth nothing briefly.

Author: Not possible as this will be part of the websocket initialising payload and that will determine the topic to which they are connected internally.

e.g.
private: false - user will be connected to a topic realtime:topic
private: true - user will be connected to a topic realtime-private:topic

    .subscribe((status: string, err: any) => {
        status == "CHANNEL_ERROR" && console.error(err.message);
    });

// Outputs the following error:
// "You do not have permissions to read from this Topic"
```

But if you connect with an authorized user you will be able to listen to all messages from the “private” topic

```tsx
// With an authenticated user
supabase.realtime.setAuth(token)

supabase
    .realtime

>     .channel("locked", { config: { private: true })

Reviewer: ```suggestion
    .channel("locked", { config: { private: true } })
```

    .subscribe((status: string, err: any) => {
        status == "CHANNEL_ERROR" && console.error(err.message);
    });
```

> ### More complex example

Reviewer: ```suggestion
### Advanced examples
```

    light:
      '/images/blog/2024-07-01-supabase-realtime-broadcast-and-presence-authorization/schema--light.png',
  }}
  alt="The table that is used to test the policies."
/>

### One user, one context, one connection

Now that we have set up everything on the database side, let's understand how it works and how we can verify authorization via RLS policies.

Upon channel subscription, we read the contents of your configuration and determine that you want to use a private channel, then we parse the information from your signed JSON Web Token (JWT), load them into a Postgres transaction using `set_config`, and verify them against the new `realtime.messages`table within the same transaction and we store the output as a group of policies within the context of the user's channel process.

<Img
  src={{
    dark: '/images/blog/2024-07-01-supabase-realtime-broadcast-and-presence-authorization/one-user-connection--dark.png',
    light:
      '/images/blog/2024-07-01-supabase-realtime-broadcast-and-presence-authorization/one-user-connection--light.png',
  }}
  alt="Flow of checks done against the databse to determine the policies for a user connection."
/>

> Since we're doing this as soon as the user subscribes to Channel we should see a small increase in subscriptions but after that we do not run any queries on the database and we use the stored context to control the actions a user can perform.

Reviewer: ```suggestion
Since we're doing this as soon as the user subscribes to Channel we should see a small latency increase in subscriptions but after that we do not run any queries on the database and we use the stored context to control the actions a user can perform.
```

I think?

description: 'Create Authorization policies for Realtime Broadcast and Presence'
author: filipe
image: 2024-07-01-supabase-realtime-broadcast-and-presence-authorization/OG.png
thumb: 2024-07-01-supabase-realtime-broadcast-and-presence-authorization/thumb.png
categories:
tags:
  - realtime
  - engineering
date: '2024-07-01'
toc_depth: 2
---

Today we're releasing Realtime Authorization for Broadcast and Presence as Public Alpha.

For context, Supabase includes three useful extensions for building real-time applications.

1. [Broadcast](https://supabase.com/docs/guides/realtime/broadcast): Send ephemeral, low-latency messages between users.
2. [Presence](https://supabase.com/docs/guides/realtime/presence): Show when users are online and share state between users.
3. [Postgres Changes](https://supabase.com/docs/guides/realtime/postgres-changes): Listen to Postgres database changes.

This release introduces authorization for Broadcast and Presence using Row Level Security policies:

```sql
CREATE POLICY "authenticated user can listen to all messages"
ON "realtime"."messages"
AS PERMISSIVE FOR SELECT
TO authenticated
USING ( true );
```

> The `realtime.messages` table is created automatically by Supabase Realtime. It represents your [Channels](/docs/guides/realtime/concepts#channels) as objects against which you can write RLS policies. We test for a user's ability to `SELECT` and `INSERT` into this table to determine whether they can join, read, or post to a Realtime Channel.

Author: Should we move the schema image we have to this section? 

cc @w3b6x9 you had a similar suggestion wdyt?

> The `realtime.messages` table is created automatically by Supabase Realtime. It represents your [Channels](/docs/guides/realtime/concepts#channels) as objects against which you can write RLS policies. We test for a user's ability to `SELECT` and `INSERT` into this table to determine whether they can join, read, or post to a Realtime Channel.

Author: Updated the document to be more inline with the feedback

This approach of using RLS policies to control access should be familiar to users of [Supabase Storage](/docs/guides/storage/security/access-control).

## How it works without Authorization

When you want to connect to a Realtime Channel, you do something close to the following code:

```tsx
// With authenticated user token
client.realtime.setAuth(token)

client.channel('channel')

> .subscribe((status: string, err: any) => {
>   status === 'SUBSCRIBED' && console.log('Connected')
> })

Reviewer 2: ```suggestion
  .subscribe((status: string, err: any) => {
    if (status === 'SUBSCRIBED') {
      console.log('Connected')
    }
  })
```

Lots of people have no idea that you can call functions like that and might be confused

Author: good call!!!

as permissive for select
to authenticated
using ( realtime.topic() = 'locked' );
```

### Enabling Authorization on Channel subscription

Now that you have your RLS policies set, you will need to define on your Channel subscription that you intend to use a private channel. To achieve it we've introduced a new configuration parameter `private` to signal to Realtime servers that you want to check authorization on the channel.

If you try to subscribe with an unauthorized user you will get a new error message informing the user that they do not have permission to access the topic.

```tsx
// With anon user
supabase.realtime
  .channel('locked', { config: { private: true } })
  .subscribe((status: string, err: any) => {
    status == 'CHANNEL_ERROR' && console.error(err.message)
  })

// Outputs the following error:
// "You do not have permissions to read from this Topic"
```

But if you connect with an authorized user you will be able to listen to all messages from the “private” topic

```tsx
// With an authenticated user
supabase.realtime.setAuth(token)

supabase.realtime
  .channel('locked', { config: { private: true } })
  .subscribe((status: string, err: any) => {

>     status == 'CHANNEL_ERROR' && console.error(err.message)

Reviewer 2: ```suggestion
    if (status === 'CHANNEL_ERROR') {
      console.error(err.message)
    }
```

Latency between geographically close users is very important for a product like Realtime. To deliver messages as fast as possible between users on our global network, we cache the policies.

We can maintain high throughput and low latency on a Realtime Channel with Broadcast and Presence authorization because:

- the policy is only generated when a user connects to a Channel
- the policy cached in memory is close to your users
- the policy is cached for the duration of the connection or until the JWT expires

Another added bonus is the fact that if a user does not have access to a given Channel they won't be able to connect at all and their connections will be rejected.

### Refreshing your Policies

Realtime already has a mechanism to check valid refreshed JWTs used by your connection as a way to ensure the user connected continues to be valid. Now, we've added a step where we re-authorize the user by querying your database's `realtime.messages` table when Realtime receives a new user token. This is to make sure the already connected user continues to be authorized even when you've updated RLS policies.

> ### Postgres Changes currently not supported

Reviewer 3: Let's rephrase this - it sounds like we don't have AuthZ at all, but what you're saying here is that it _is_ supported in a different way

Author: changed it slightly on the header and changed the section below it

Realtime Authorization currently supports only Broadcast and Presence. Postgres Changes already adheres to RLS policies on the tables you're listening to so you can continue using that authorization scheme for getting changes from your database.

## Availability

> Broadcast and Presence Authorization is available in Public Alpha. We are looking for feedback so please do share it in the [Github discussion](github.com/orgs/supabase/discussions/22484).

Reviewer 3: GitHub with a capital "H"
